crates/net/downloaders/src/bitfinity_evm_client.rs

```diff
@@ -73,10 +92,179 @@ pub struct CertificateCheckSettings {
     pub ic_root_key: String,
 }
 
+/// RPC client configuration
+#[derive(Debug, Clone)]
+pub struct RpcClientConfig {
+    /// Primary RPC URL
+    pub primary_url: String,
+    /// Backup RPC URL
+    pub backup_url: Option<String>,
+    /// Maximum number of retries
+    pub max_retries: u32,
+    /// Delay between retries
+    pub retry_delay: Duration,
+}
+
 impl BitfinityEvmClient {
+    /// Creates a new RPC client for the given URL
+    fn create_client(url: &str) -> EthJsonRpcClient<ReqwestClient> {
+        EthJsonRpcClient::new(ReqwestClient::new(url.to_string()))
+    }
+
+    /// Create a new `BitfinityEvmClient` from RPC configuration
+    pub fn new(config: RpcClientConfig) -> Self {
+        let client = Self::create_client(&config.primary_url);
+
+        Self {
+            rpc_config: config,
+            client: RwLock::new(client),
+            using_backup: Mutex::new(false),
+            headers: HashMap::new(),
+            hash_to_number: HashMap::new(),
+            bodies: HashMap::new(),
+        }
+    }
+
+    /// Try to switch back to primary URL if we're using backup
+    async fn try_primary_url(&self) -> Result<(), RemoteClientError> {
+        // Early return if not using backup
+        {
+            let backup_status = self.using_backup.lock();
+            if !*backup_status {
+                return Ok(());
+            }
+        }
+        debug!(target: "downloaders::bitfinity_evm_client", "Attempting to switch back to primary URL");
+
+        // Create new client and test connection
+        let new_client = Self::create_client(&self.rpc_config.primary_url);
+        match new_client.get_block_number().await {
+            Ok(_) => {
+                // Update client and status atomically
+                let mut client = self.client.try_write().ok_or_else(|| {
+                    RemoteClientError::InternalError("Failed to acquire `client` lock".to_string())
+                })?;
+
+                let mut using_backup = self.using_backup.try_lock().ok_or_else(|| {
+                    RemoteClientError::InternalError(
+                        "Failed to acquire `using_backup` lock".to_string(),
+                    )
+                })?;
+
+                *client = new_client;
+                *using_backup = false;
+                info!(target: "downloaders::bitfinity_evm_client", "Successfully switched back to primary URL");
+                Ok(())
+            }
+            Err(e) => {
+                warn!(target: "downloaders::bitfinity_evm_client", "Failed to switch back to primary URL: {}, Primary URL still unavailable", e);
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Switch to backup URL if available/// Switch to backup URL if available
+    async fn switch_to_backup(&self) -> Result<(), RemoteClientError> {
```

Comment:
Same as before, this is overcomplex. We should use the backup URL in case the first one fails